File: src/moonlight/runner.py
"""Scenario runner — loads YAML, drives Playwright, calls vision checks, writes report."""
from __future__ import annotations

import asyncio
import json
import re
import shutil
from dataclasses import dataclass, field, asdict
from datetime import datetime
from pathlib import Path
from typing import Any

import yaml


# Patterns produced by the OLD regex-based recorder. The selectors are mangled
# in two ways depending on what the user recorded:
#   1. Plain role+name:        textbox", name="Username
#   2. YAML-escaped role+name: button", name="Login 
#   3. Quoted (rare):          "button", name="Yes"
# All collapse to <role>", name="<name>  after stripping any stray quote chars.
_BROKEN_SELECTOR_RE = re.compile(
    r'^"?(?P<role>[A-Za-z_][A-Za-z0-9_-]*)"?\s*,\s*name="?(?P<name>.*?)"?\s*$'
)


def _migrate_step(step: dict) -> dict:
    """Repair locators recorded by the pre-AST parser. No-op on clean YAMLs."""
    sel = step.get("selector")
    if isinstance(sel, str) and "name=" in sel:
        m = _BROKEN_SELECTOR_RE.match(sel.strip())
        if m:
            step.pop("selector")
            step["role"] = m.group("role")
            name = m.group("name")
            if name:
                step["name"] = name
    return step

from .assertions import CheckResult, ai_check, ocr_check
from .browser import Browser
from .diff import pixel_diff
from .vision import SmolVLM2Client


# ---------- data classes ----------------------------------------------------

@dataclass
class StepResult:
    index: int
    action: str
    name: str
    passed: bool
    checks: list[dict[str, Any]] = field(default_factory=list)
    screenshot: str | None = None
    baseline: str | None = None
    diff_overlay: str | None = None
    diff_ratio: float | None = None
    error: str | None = None
    ai_narration: str | None = None
    console_messages: list[dict] = field(default_factory=list)
    failed_requests: list[dict] = field(default_factory=list)
    page_errors: list[str] = field(default_factory=list)


@dataclass
class RunResult:
    scenario: str
    url: str
    started_at: str
    finished_at: str
    passed: bool
    steps: list[StepResult]
    run_dir: str
    ai_summary: str | None = None


# ---------- helpers ---------------------------------------------------------

def _load_config(project_root: Path) -> dict:
    return yaml.safe_load((project_root / "config.yaml").read_text())


def _load_scenario(scenario_path: Path) -> dict:
    return yaml.safe_load(scenario_path.read_text())


def _slug(name: str) -> str:
    return "".join(c if c.isalnum() or c in "-_" else "-" for c in name).strip("-")


# ---------- single-step dispatcher -----------------------------------------

async def _do_step(
    idx: int,
    step: dict,
    browser: Browser,
    vision: SmolVLM2Client | None,
    run_dir: Path,
    baseline_dir: Path,
    diff_threshold: float,
    mode: str,  # "baseline" or "run"
) -> StepResult:
    step = _migrate_step(step)
    action = step["action"]
    # `name` is the screenshot label, but `name=` can also be a locator kwarg
    # for get_by_role. For step naming, prefer an explicit "label" field then
    # fall back to action+idx so role-name doesn't leak into the screenshot id.
    if action == "screenshot":
        name = step.get("name") or f"{action}-{idx:02d}"
    else:
        name = f"{action}-{idx:02d}"
    result = StepResult(index=idx, action=action, name=name, passed=True)

    try:
        if action == "goto":
            await browser.goto(step["url"])
        elif action == "click":
            await browser.do_click(step)
        elif action == "fill":
            await browser.do_fill(step)
        elif action == "select":
            await browser.do_select(step)
        elif action == "press":
            await browser.do_press(step)
        elif action == "wait_for":
            await browser.do_wait_for(step)
        elif action == "wait_ms":
            await browser.wait_ms(int(step["ms"]))
        elif action == "scroll":
            await browser.scroll(int(step.get("y", 400)))
        elif action == "screenshot":
            shot = run_dir / "screenshots" / f"{idx:02d}-{_slug(name)}.png"
            await browser.screenshot(shot, full_page=bool(step.get("full_page")))
            result.screenshot = str(shot)

            # baseline mode → save and skip checks
            if mode == "baseline":
                base_target = baseline_dir / f"{idx:02d}-{_slug(name)}.png"
                base_target.parent.mkdir(parents=True, exist_ok=True)
                shutil.copyfile(shot, base_target)
                result.baseline = str(base_target)
                return result

            # run mode → regression diff + custom checks
            baseline_path = baseline_dir / f"{idx:02d}-{_slug(name)}.png"
            if baseline_path.exists():
                overlay_path = run_dir / "diffs" / f"{idx:02d}-{_slug(name)}.png"
                diff = pixel_diff(baseline_path, shot, overlay_path)
                result.baseline = str(baseline_path)
                result.diff_overlay = str(diff.overlay_path)
                result.diff_ratio = diff.ratio

                if diff.ratio > diff_threshold and vision is not None:
                    verdict = vision.regression_judge(baseline_path, shot)
                    result.checks.append({
                        "kind": "regression_judge",
                        "passed": verdict.passed,
                        "detail": f"diff_ratio={diff.ratio:.4f} | model: {verdict.raw}",
                    })
                    if not verdict.passed:
                        result.passed = False

            if vision is not None:
                # AI narration runs on EVERY screenshot — single sentence for the report.
                try:
                    result.ai_narration = vision.narrate(shot)
                except Exception as e:
                    result.ai_narration = f"(narration unavailable: {e})"

                if (ac := step.get("ai_check")):
                    if "TODO" not in str(ac.get("prompt", "")).upper():
                        cr = ai_check(vision, shot, ac["prompt"], ac.get("expect", "yes"))
                        result.checks.append(asdict(cr))
                        if not cr.passed:
                            result.passed = False
                if (oc := step.get("ocr_check")):
                    cr = ocr_check(vision, shot, list(oc.get("contains", [])))
                    result.checks.append(asdict(cr))
                    if not cr.passed:
                        result.passed = False
        else:
            raise ValueError(f"unknown action: {action}")
    except Exception as e:
        result.passed = False
        result.error = f"{type(e).__name__}: {e}"

    # Capture browser events that fired during this step.
    events = browser.snapshot_events()
    result.console_messages = events["console"]
    result.failed_requests = events["failed_requests"]
    result.page_errors = events["page_errors"]
    return result


# ---------- public API ------------------------------------------------------

async def _execute(
    scenario: dict,
    project_root: Path,
    mode: str,
    on_log=None,
) -> RunResult:
    cfg = _load_config(project_root)
    scenario_name = scenario["name"]
    started = datetime.now()
    stamp = started.strftime("%Y%m%d-%H%M%S")
    run_dir = project_root / cfg["paths"]["runs"] / f"{stamp}-{scenario_name}"
    (run_dir / "screenshots").mkdir(parents=True, exist_ok=True)
    (run_dir / "diffs").mkdir(parents=True, exist_ok=True)
    baseline_dir = project_root / cfg["paths"]["baselines"] / scenario_name

    vision = None
    if mode == "run":
        vision = SmolVLM2Client(
            host=cfg["ollama"]["host"],
            model=cfg["ollama"]["model"],
            timeout_s=cfg["ollama"]["timeout_s"],
            options=cfg["ollama"].get("options"),
        )

    viewport = scenario.get("viewport") or cfg["browser"]["viewport"]
    headless = scenario.get("headless", cfg["browser"]["headless"])
    threshold = float(cfg["diff"]["threshold"])
    storage_state = scenario.get("storage_state")
    if storage_state:
        storage_state = project_root / storage_state

    def _log(line: str) -> None:
        if on_log:
            on_log(line)

    _log(f"[start] mode={mode} url={scenario['url']} steps={len(scenario['steps'])}")
    steps: list[StepResult] = []
    async with Browser(
        base_url=scenario["url"],
        viewport=viewport,
        headless=headless,
        default_timeout_ms=cfg["browser"]["default_timeout_ms"],
        storage_state=storage_state,
    ) as browser:
        for idx, step in enumerate(scenario["steps"], start=1):
            preview = step.get("url") or step.get("selector") or step.get("name") or ""
            _log(f"[step {idx:02d}] {step['action']} {preview}")
            res = await _do_step(idx, step, browser, vision, run_dir, baseline_dir, threshold, mode)
            steps.append(res)
            verdict = "PASS" if res.passed else "FAIL"
            if res.ai_narration:
                _log(f"   ↳ AI sees: {res.ai_narration}")
            for c in res.checks:
                _log(f"   ↳ AI {c['kind']}: {'PASS' if c['passed'] else 'FAIL'}")
            if res.failed_requests:
                _log(f"   ↳ network: {len(res.failed_requests)} failed request(s)")
            if res.page_errors:
                _log(f"   ↳ JS errors: {len(res.page_errors)}")
            if res.error:
                _log(f"   ↳ error: {res.error}")
            _log(f"   ↳ verdict: {verdict}")

    finished = datetime.now()
    overall = all(s.passed for s in steps)

    # AI run summary — only in 'run' mode (baseline has no verdicts to summarize).
    summary: str | None = None
    if mode == "run" and vision is not None:
        log_lines = []
        for s in steps:
            line = f"Step {s.index} [{s.action}] {'PASS' if s.passed else 'FAIL'}"
            if s.ai_narration:
                line += f" — {s.ai_narration}"
            if s.error:
                line += f" — error: {s.error[:120]}"
            for c in s.checks:
                if not c.get("passed"):
                    line += f" — AI {c['kind']}: {c['detail'][:120]}"
            log_lines.append(line)
        try:
            summary = vision.summarize_run("\n".join(log_lines))
            _log(f"[summary] {summary}")
        except Exception as e:
            summary = f"(summary unavailable: {e})"

    return RunResult(
        scenario=scenario_name,
        url=scenario["url"],
        started_at=started.isoformat(timespec="seconds"),
        finished_at=finished.isoformat(timespec="seconds"),
        passed=overall,
        steps=steps,
        run_dir=str(run_dir),
        ai_summary=summary,
    )


def run_scenario(scenario_path: Path, project_root: Path, mode: str = "run", on_log=None) -> RunResult:
    scenario = _load_scenario(scenario_path)
    return asyncio.run(_execute(scenario, project_root, mode, on_log=on_log))


def write_json(result: RunResult, out_path: Path) -> None:
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(asdict(result), indent=2))

File: internal/runner/runner.go
// Package runner executes a scenario start-to-finish: drives the browser
// step by step, runs DOM and (optionally) AI assertions on screenshots, and
// returns a RunResult.
//
// Modes:
//   - ModeBaseline → screenshots are copied into baselineDir; no assertions run.
//   - ModeRun      → screenshots are compared against baseline (pixel diff),
//                    DOM assertions evaluated, AI assertions evaluated IF the
//                    scenario carries any ai_check and the vision client is
//                    available.
package runner

import (
	"fmt"
	"path/filepath"
	"strings"
	"time"

	"github.com/krisk248/moonlight/internal/browser"
	"github.com/krisk248/moonlight/internal/diff"
	"github.com/krisk248/moonlight/internal/dom"
	"github.com/krisk248/moonlight/internal/scenario"
	"github.com/krisk248/moonlight/internal/vision"
)

type Mode string

const (
	ModeBaseline Mode = "baseline"
	ModeRun      Mode = "run"
)

// StepResult is the outcome of one step.
type StepResult struct {
	Index          int                    `json:"index"`
	Action         string                 `json:"action"`
	Name           string                 `json:"name"`
	Passed         bool                   `json:"passed"`
	Error          string                 `json:"error,omitempty"`
	ScreenshotPath string                 `json:"screenshot,omitempty"`
	BaselinePath   string                 `json:"baseline,omitempty"`
	DiffOverlay    string                 `json:"diff_overlay,omitempty"`
	DiffRatio      float64                `json:"diff_ratio,omitempty"`
	AINarration    string                 `json:"ai_narration,omitempty"`
	Checks         []CheckResult          `json:"checks,omitempty"`
	Console        []browser.ConsoleEntry `json:"console_messages,omitempty"`
	Network        []browser.NetworkEntry `json:"failed_requests,omitempty"`
	PageErrors     []string               `json:"page_errors,omitempty"`
}

type CheckResult struct {
	Kind   string `json:"kind"`
	Passed bool   `json:"passed"`
	Detail string `json:"detail"`
}

// RunResult is the outcome of a whole scenario run.
type RunResult struct {
	Scenario   string       `json:"scenario"`
	URL        string       `json:"url"`
	StartedAt  string       `json:"started_at"`
	FinishedAt string       `json:"finished_at"`
	Passed     bool         `json:"passed"`
	Mode       string       `json:"mode"`
	AIUsed     bool         `json:"ai_used"`
	Steps      []StepResult `json:"steps"`
	RunDir     string       `json:"run_dir"`
	AISummary  string       `json:"ai_summary,omitempty"`
}

// Opts groups dependencies passed into Run.
type Opts struct {
	ProjectRoot       string
	BaselineDir       string
	RunsDir           string
	Vision            *vision.Client // may be nil; consulted only if scenario needs AI
	Headless          bool
	DiffTolerance     uint8 // 0..255 channel-delta threshold per pixel; 12 is a reasonable default
	DefaultTimeoutMS  int   // global Playwright timeout per action; 0 → 15000
	OnLog             func(string)
}

// VisionClient narrows the vision.Client API to what the runner needs so
// tests can pass a mock. (Currently unused in tests; exported for future use.)
type VisionClient interface {
	FunctionalCheck(imagePath, question string, expectYes bool) (vision.Verdict, error)
	Narrate(imagePath string) (string, error)
	SummarizeRun(runLog string) (string, error)
	Ping() bool
}

func slug(s string) string {
	var b strings.Builder
	for _, r := range s {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9', r == '-', r == '_':
			b.WriteRune(r)
		default:
			b.WriteRune('-')
		}
	}
	return strings.Trim(b.String(), "-")
}

// Run executes a scenario.
func Run(s *scenario.Scenario, mode Mode, opts Opts) (*RunResult, error) {
	if opts.DiffTolerance == 0 {
		opts.DiffTolerance = 12
	}

	started := time.Now().UTC()
	stamp := started.Format("20060102-150405")
	runDir := filepath.Join(opts.RunsDir, fmt.Sprintf("%s-%s", stamp, s.Name))
	baselineDir := filepath.Join(opts.BaselineDir, s.Name)

	log := opts.OnLog
	if log == nil {
		log = func(string) {}
	}

	// Per-scenario decision: do we even talk to the AI?
	useAI := mode == ModeRun && s.NeedsAI() && opts.Vision != nil && opts.Vision.Ping()
	if mode == ModeRun && s.NeedsAI() && !useAI {
		log("[ai] scenario requests ai_check but vision is unavailable — skipping AI assertions")
	}

	log(fmt.Sprintf("[start] mode=%s url=%s steps=%d ai=%v", mode, s.URL, len(s.Steps), useAI))

	headless := opts.Headless
	if s.Headless != nil {
		headless = *s.Headless
	}

	storageStatePath := ""
	if s.StorageState != "" {
		storageStatePath = filepath.Join(opts.ProjectRoot, s.StorageState)
	}

	timeoutMS := opts.DefaultTimeoutMS
	if timeoutMS <= 0 {
		timeoutMS = 15000
	}
	br, err := browser.New(browser.Opts{
		BaseURL:          s.URL,
		Viewport:         s.Viewport,
		Headless:         headless,
		DefaultTimeoutMS: timeoutMS,
		StorageStatePath: storageStatePath,
	})
	if err != nil {
		return nil, err
	}
	defer br.Close()

	var steps []StepResult
	for i, step := range s.Steps {
		idx := i + 1
		name := step.Name
		if step.Action != "screenshot" || name == "" {
			name = fmt.Sprintf("%s-%02d", step.Action, idx)
		}
		log(fmt.Sprintf("[step %02d] %s", idx, step.Action))

		res := StepResult{Index: idx, Action: step.Action, Name: name, Passed: true}

		// Per-step timeout override — restored after the action.
		if step.TimeoutMS > 0 {
			br.SetStepTimeout(step.TimeoutMS)
		}

		var stepErr error
		switch step.Action {
		case "goto":
			stepErr = br.Goto(step.URL)
		case "click":
			stepErr = br.Click(step)
		case "fill":
			stepErr = br.Fill(step)
		case "select":
			stepErr = br.SelectOption(step)
		case "press":
			stepErr = br.Press(step)
		case "wait_for":
			stepErr = br.WaitFor(step)
		case "wait_ms":
			br.WaitMS(step.MS)
		case "wait_for_networkidle":
			stepErr = br.WaitForNetworkidle()
		case "upload_file":
			stepErr = br.Upload(step)
		case "download_file":
			stepErr = br.Download(step)
		case "scroll":
			stepErr = br.Scroll(step.Y)
		case "screenshot":
			shot := filepath.Join(runDir, "screenshots", fmt.Sprintf("%02d-%s.png", idx, slug(name)))
			stepErr = br.Screenshot(shot, step.FullPage)
			if stepErr == nil {
				res.ScreenshotPath = shot
				if mode == ModeBaseline {
					base := filepath.Join(baselineDir, fmt.Sprintf("%02d-%s.png", idx, slug(name)))
					if cperr := copyFile(shot, base); cperr == nil {
						res.BaselinePath = base
					}
				} else {
					// pixel diff against baseline
					base := filepath.Join(baselineDir, fmt.Sprintf("%02d-%s.png", idx, slug(name)))
					if fileExists(base) {
						overlay := filepath.Join(runDir, "diffs", fmt.Sprintf("%02d-%s.png", idx, slug(name)))
						if d, derr := diff.Diff(base, shot, overlay, opts.DiffTolerance); derr == nil {
							res.BaselinePath = base
							res.DiffOverlay = overlay
							res.DiffRatio = d.Ratio
						}
					}

					// dom_check (deterministic, always runs)
					if step.DOMCheck != nil {
						domRes := dom.Evaluate(br.Page(), step.DOMCheck)
						cr := CheckResult{
							Kind:   "dom_check",
							Passed: domRes.Passed,
							Detail: strings.Join(domRes.Details, " | "),
						}
						res.Checks = append(res.Checks, cr)
						if !cr.Passed {
							res.Passed = false
						}
					}

					// ai_check (optional)
					if useAI && step.AICheck != nil {
						q := step.AICheck.Prompt
						if !strings.Contains(strings.ToUpper(q), "TODO") {
							expectYes := strings.EqualFold(step.AICheck.Expect, "yes") ||
								strings.EqualFold(step.AICheck.Expect, "true")
							v, verr := opts.Vision.FunctionalCheck(shot, q, expectYes)
							cr := CheckResult{
								Kind:   "ai_check",
								Passed: v.Passed,
								Detail: fmt.Sprintf("expected=%s | model: %s", yesOrNo(expectYes), v.Raw),
							}
							if verr != nil {
								cr.Detail = "AI error: " + verr.Error()
							}
							res.Checks = append(res.Checks, cr)
							if !cr.Passed {
								res.Passed = false
							}
						}
						if narr, nerr := opts.Vision.Narrate(shot); nerr == nil {
							res.AINarration = narr
						}
					}
				}
			}
		default:
			stepErr = fmt.Errorf("unknown action: %s", step.Action)
		}
		if stepErr != nil {
			res.Passed = false
			res.Error = stepErr.Error()
		}

		// Restore the global timeout for the next step.
		if step.TimeoutMS > 0 {
			br.ResetTimeout(timeoutMS)
		}

		cons, net, errs := br.SnapshotEvents()
		res.Console = cons
		res.Network = net
		res.PageErrors = errs
		steps = append(steps, res)
	}

	overall := true
	for _, st := range steps {
		if !st.Passed {
			overall = false
			break
		}
	}

	rr := &RunResult{
		Scenario:   s.Name,
		URL:        s.URL,
		StartedAt:  started.Format(time.RFC3339),
		FinishedAt: time.Now().UTC().Format(time.RFC3339),
		Passed:     overall,
		Mode:       string(mode),
		AIUsed:     useAI,
		Steps:      steps,
		RunDir:     runDir,
	}

	if useAI {
		summary, _ := opts.Vision.SummarizeRun(buildSummaryLog(steps))
		rr.AISummary = summary
	}
	return rr, nil
}

func yesOrNo(b bool) string {
	if b {
		return "YES"
	}
	return "NO"
}

func buildSummaryLog(steps []StepResult) string {
	var b strings.Builder
	for _, s := range steps {
		v := "PASS"
		if !s.Passed {
			v = "FAIL"
		}
		fmt.Fprintf(&b, "Step %d [%s] %s", s.Index, s.Action, v)
		if s.AINarration != "" {
			fmt.Fprintf(&b, " — %s", s.AINarration)
		}
		if s.Error != "" {
			fmt.Fprintf(&b, " — error: %s", s.Error)
		}
		b.WriteByte('\n')
	}
	return b.String()
}

func fileExists(path string) bool {
	_, err := osStat(path)
	return err == nil
}
